fix: Evaluate boolean values better via /api/resource/<doctype>

gavindsouza opened the pull request:
`GET /api/resource/ToDo?limit=10&debug=False&as_dict=0` would be received by the resource handler as debug="False" and as_dict="0" which are both truthy values.

So, even though you requested for a list of lists response without debugging on, you'd get the exact opposite; debug on and a list of dicts.

- Evaluate boolean values for `GET /api/resource/<doctype>`
- Added `limit` parameter as an alias for `limit_page_length`
- Added `frappe.utils.data.sbool` that converts strings to bool values if applicable.
- Added some seemingly stupid comments for the sake of consistency.

**Pending:**
- [x] Add tests
- [x] Update docs


Fixes https://github.com/frappe/frappe/issues/13117

gavindsouza committed:
fix: Evaluate boolean values better via /api/resource/<doctype>

`GET /api/resource/ToDo?limit=10&debug=False&as_dict=0` would be received by the resource handler as debug="False" and as_dict="0" which are both truthy values.

So, even though you requested for a list of lists response without debugging on, you'd get the exact opposite; debug on and a list of dicts.

- Evaluate boolean values for `GET /api/resource/<doctype>`
- Added `limit` parameter as an alias for `limit_page_length`
- Added `frappe.utils.data.sbool` that converts strings to bool values if applicable.
- Added some seemingly stupid comments for the sake of consistency.

gavindsouza committed:
chore: Renamed test files "appropriately"

gavindsouza committed:
chore: Rename test suite and remove unused imports

gavindsouza committed:
Merge branch 'develop' into api-updates

gavindsouza committed:
test: Added tests for Frappe APIs

* Added tests for /api/resource usage
* Added tests for /api/method base endpoints

gavindsouza committed:
fix(test): Set admin password correctly

gavindsouza commented:
@Mergifyio backport version-13-hotfix

gavindsouza commented:
Docs updated via https://github.com/frappe/frappe_docs/pull/133

hasnain2808 commented on frappe/api.py:
```diff
@@ -108,25 +109,39 @@ def handle():
+
+					# convert strings to native types
+					frappe.local.form_dict.update(
+						{x: sbool(y) for x, y in frappe.local.form_dict.items()}
```
this would convert in all these cases

![image](https://user-images.githubusercontent.com/28212972/119487423-fb9b9480-bd76-11eb-9b62-dda6068bd695.png)


This would be applicable to all parameters regardless of what was expected

One idea that I have is that can we keep the util and apply it to a parameter if we are expecting a boolean value at specific endpoint ?

gavindsouza replied:
Given the scope of its usage at this point, this becomes a problem when you'd have a field named `y`,`n`, `true`, `false` and `order_by` that field, or have the same values for a `document name` that `parent` parameter would accept out of all those that Frappe REST allows.

OTOH, since only debug and as_dict can have boolean values, might as well just convert them 😄 . Nice catch

gavindsouza committed:
fix: Convert only as_dict and debug values to bool

Given the scope of its usage at this point, this becomes a problem when you'd have a field named y,n, true, false and order_by that field, or have the same values for a document name that parent parameter would accept out of all those that Frappe REST allows.
